### DISCUSSION TITLE
Disabling a service and post-rules in configlets

### DISCUSSION THREAD

**Discussion by Hexxel** [finit-project/finit#399] [Q&A]
If a service is disabled and the service has a post-rule in it's configlet, the post-rule doesn't seems to run. Is there a preferred way of handling this or should finit go from enabled -> stopped -> disabled in this case?

**Comment by troglobit** [comment]
So, just to clarify, the service is running when it's then disabled and `initctl reload` is called?

The latter is the [intended state transition](https://github.com/troglobit/finit/blob/master/doc/service.md#state-machine): RUNNING -> -E -> STOPPING -> post:script -> CLEANUP -> HALTED.

----

The `pre-post-serv.sh` test does not check for disable today, only stop.  It would be great to reproduce this issue in that controlled environment.  Unfortunately I'm a bit busy next next couple of days, but if you like, you could give it a shot: add a stage to the `test_one()` function that also verifies disabling a service to trigger the post script.

**Tip:** I use the `./check.sh` script in the root directory to configure finit and run the tests.  To run a single test after that, either:

  - `make check TESTS=pre-post-serv.sh`
  - or `cd test/; ./pre-post-serv.sh` for an interactive run

**Comment by Hexxel** [comment]
Is there a docker image I could use to run the tests?

**Reply by troglobit** [reply]
Unfortunately not.  A PC with the latest Ubuntu LTS, or equivalent Debian or Linux Mint, should be sufficient though.  Both libite and libuev are available in Debian/Ubuntu these days, so you can:

    sudo apt install libite-dev libuev-dev

in addition to the standard packages when building autotools software from GIT:

    sudo apt install autoconf automake libtool pkg-config

The test harness itself runs the test in a chroot+unshare with a static BusyBox (x86_64) that it downloads from the Internet and uses as a root filesystem with Finit on top.  All tests run from inside that isolated namespace using pure POSIX shell scripts.

... or wait for me to get my nose above water and have a look at it myself :smiley:

**Comment by troglobit** [comment]
I've now updated the `pre-post-serv.sh` regression test to include a second pass for services run from configlets in `/etc/finit.d/available -> ../enabled`.  See change in eb4eca96.

See the latest run here: https://github.com/troglobit/finit/actions/runs/8400401702

As you can see, the post:script is called in all permutations the test runs.  That's not to say there *isn't* a possible bug here, only that I cannot reproduce it.  If you can describe or provide a test to reproduce the issues you see, with the latest master, I'm more than happy to have a look at it.

Many stability fixes have gone into Finit over the last couple of releases.  If you are not running the latest v4.7 release, I suggest planning an upgrade.  There are also a few interesting fixes on `master` that have not yet been released.  If time permits I'll push out a v4.8 in the next 30 days, but things are hectic at $DAYJOB atm, lots of work for customers before end of Q2, so likely the next Finit release will be during my vacation.

**Comment by Hexxel** [comment]
Ok, many thanks. I'll go back to see if our handling of states are ok.
